// client/src/hooks/useAuth.tsx
import { createContext, useContext, useState, useEffect, useMemo } from 'react';
import type { ReactNode } from 'react';
import api from '../lib/api';
import { useNavigate } from 'react-router-dom';

interface User {
  id: number;
  email: string;
  username: string;
  avatar_url?: string;
}

interface AuthContextType {
  user: User | null;
  login: (email: string, password: string) => Promise<void>;
  register: (email: string, password: string, username: string) => Promise<void>;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | null>(null);

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const navigate = useNavigate();

  useEffect(() => {
    const token = localStorage.getItem('token');
    if (token) {
      // Configure axios avec le token
  api.defaults.headers.common['Authorization'] = `Bearer ${token}`;
      // Récupérer les infos utilisateur
      void fetchUserInfo();
    }
  }, []);

  const fetchUserInfo = async () => {
    try {
      const token = localStorage.getItem('token');
      if (!token) return;

  const { data } = await api.get('/auth/me');
      setUser(data);
    } catch {
      logout();
    }
  };

  const login = async (email: string, password: string) => {
    try {
  const { data } = await api.post('/auth/login', {
        email,
        password,
      });

      localStorage.setItem('token', data.token);
  api.defaults.headers.common['Authorization'] = `Bearer ${data.token}`;
      setUser(data.user);
    } catch {
      throw new Error('Invalid credentials');
    }
  };

  const register = async (email: string, password: string, username: string) => {
    try {
  const { data } = await api.post('/auth/register', {
        email,
        password,
        username,
      });

      localStorage.setItem('token', data.token);
  api.defaults.headers.common['Authorization'] = `Bearer ${data.token}`;
      setUser(data.user);
    } catch {
      throw new Error('Registration failed');
    }
  };

  const logout = () => {
    localStorage.removeItem('token');
  delete api.defaults.headers.common['Authorization'];
    setUser(null);
    navigate('/');
  };

  const value = useMemo(
    () => ({ user, login, register, logout }),
    [user, navigate]
  );

  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
